### DISCUSSION TITLE
Error: Calling setProviderByEnv or setProviderByName removes this capability.

### DISCUSSION THREAD

**Discussion by Fraccas** [reach-sh/reach-lang#1004] [Q&A] [answered]
Users testing my system are occasionally getting the following error message: 
`Error: Providers created by environment do not have default addresses. Calling setProviderByEnv or setProviderByName removes this capability. Try removing calls to those functions.`

I've seen this happen when you use 
`reach.setProviderByName('TestNet'); `
combined with
`this.acc = await this.reach.getDefaultAccount();`
but in my case I'm not using "setProviderByName()" in the frontend

My frontend code 
```
    this.reach.setWalletFallback(this.reach.walletFallback({
      providerEnv: {
        ALGO_TOKEN: '',
        ALGO_SERVER: "https://testnet-api.algonode.cloud",
        ALGO_PORT: '',
        ALGO_INDEXER_TOKEN: '',
        ALGO_INDEXER_SERVER: "https://testnet-idx.algonode.cloud",
        ALGO_INDEXER_PORT: '',
      }, MyAlgoConnect}));
 ```

`this.acc = await this.reach.getDefaultAccount();`

Any thoughts on how this could be occurring? It's somewhat rare and is usually resolved with a page refresh. This is an Angular site. 

Note: I do use `reach.setProviderByName('TestNet');` within the API that this site connects to, and the backend uses `const acc = await reach.newAccountFromMnemonic(process.env.CreatorSK); ` to pull the account.

Note2: This seems to only occur on mobile so far. 

![image](https://user-images.githubusercontent.com/3107701/168967568-e36efe26-51b9-407c-9745-5f67e578d70f.png)

**Comment by jeapostrophe** [comment, marked as answer]
My guess is that there is a call to a Reach function before you set the wallet fallback somehow and so it is using the default provider, which has this error.

**Reply by Fraccas** [reply]
> My guess is that there is a call to a Reach function before you set the wallet fallback somehow and so it is using the default provider, which has this error.

That may be possible. The following coding is in the Angular ngInit()
```
this.reach.setWalletFallback(this.reach.walletFallback({
      providerEnv: {
        ALGO_TOKEN: '',
        ALGO_SERVER: "https://testnet-api.algonode.cloud",
        ALGO_PORT: '',
        ALGO_INDEXER_TOKEN: '',
        ALGO_INDEXER_SERVER: "https://testnet-idx.algonode.cloud",
        ALGO_INDEXER_PORT: '',
      }, MyAlgoConnect}));
```

The `this.acc = await this.reach.getDefaultAccount();` code is located in a button click function. 

```
 MyAlgoWalletConnect = async() => {
    this.modalMsg = "Please connect wallet...";
    this.open(this.loadingModal);
    try {
      this.acc = await this.reach.getDefaultAccount();
      this.wallet_address = this.reach.formatAddress(await this.acc.getAddress());
      const balAtomic = await this.reach.balanceOf(this.acc);
      this.balance = this.reach.formatCurrency(balAtomic, 4);
      this.balanceFloat = parseFloat(this.balance);
      this.isLogged = true;
    } catch (e) {
      document.getElementById("closeLoadingModal").click();
      //this.errorMsg = this.popupError;
      this.errorMsg = e;
      this.open(this.errorModal);
    }

    document.getElementById("closeLoadingModal").click();

    await this.CheckRequiredOptin();
  }
```

The following code has been reduced to only what is relevant to this topic. 
```
  async ngOnInit() {
    
    this.reach.setWalletFallback(this.reach.walletFallback({
      providerEnv: {
        ALGO_TOKEN: '',
        ALGO_SERVER: "https://testnet-api.algonode.cloud",
        ALGO_PORT: '',
        ALGO_INDEXER_TOKEN: '',
        ALGO_INDEXER_SERVER: "https://testnet-idx.algonode.cloud",
        ALGO_INDEXER_PORT: '',
      }, MyAlgoConnect}));

    await this.RequestShufflePool();
    this.hasLoaded = true;
  }
```

I've added a hasLoaded check after an awaited function to show the "Connect Wallet" button conditionally, but as setWalletFallback doesn't look to be awaitable I'm not sure if this really helps.

**Reply by jeapostrophe** [reply]
I think you should `setWalletFallback` immediately
